src/lib.rs

```diff
@@ -1,101 +1,405 @@
 pub use rule_tester::{RuleTestInvalid, RuleTester, RuleTests};
-use tree_sitter::{Node, Query};
-use violation::ViolationBuilder;
+use tree_sitter::Query;
+use violation::{ViolationBuilder, ViolationWithContext};
 
-use crate::context::Context;
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
 
-#[macro_export]
-macro_rules! regex {
-    ($re:expr $(,)?) => {{
-        static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
-        RE.get_or_init(|| regex::Regex::new($re).unwrap())
-    }};
+pub fn run_and_output(config: Config) {
+    let violations = run(config);
+    if violations.is_empty() {
+        process::exit(0);
+    }
+    for violation in violations {
+        violation.print();
+    }
+    process::exit(1);
 }
 
-const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
-const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
+const MAX_FIX_ITERATIONS: usize = 10;
+
+pub fn run(config: Config) -> Vec<ViolationWithContext> {
+    let resolved_rules = config.get_resolved_rules();
+    let aggregated_queries = AggregatedQueries::new(&resolved_rules, &config);
+    let tree_sitter_grep_args = get_tree_sitter_grep_args(&aggregated_queries);
+    let all_violations: Mutex<HashMap<PathBuf, Vec<ViolationWithContext>>> = Default::default();
+    let files_with_fixes: AllPendingFixes = Default::default();
+    tree_sitter_grep::run_with_callback(
+        tree_sitter_grep_args.clone(),
+        |capture_info, file_contents, path| {
+            let (rule, rule_listener) =
+                aggregated_queries.get_rule_and_listener(capture_info.pattern_index);
+            let mut query_match_context =
+                QueryMatchContext::new(path, file_contents, rule, &config);
+            (rule_listener.on_query_match)(capture_info.node, &mut query_match_context);
+            if let Some(violations) = query_match_context.violations.take() {
+                all_violations
+                    .lock()
+                    .unwrap()
+                    .entry(path.to_owned())
+                    .or_default()
+                    .extend(violations);
+            }
+            if let Some(fixes) = query_match_context.into_pending_fixes() {
+                assert!(config.fix);
+                files_with_fixes.append(path, file_contents, &rule.meta.name, fixes);
+            }
+        },
+    )
+    .unwrap();
+    let mut all_violations = all_violations.into_inner().unwrap();
+    if !config.fix {
+        return all_violations.into_values().flatten().collect();
+    }
+    let files_with_fixes = files_with_fixes.into_inner().unwrap();
+    if files_with_fixes.is_empty() {
+        return all_violations.into_values().flatten().collect();
+    }
+    let aggregated_results_from_files_with_fixes: HashMap<
```

Comment:
Then if we do need to go into "fixing mode", the way that felt like it wanted to be structured was "independent fixing loops for each file that reported fixes"

And so that felt parallelizable so used `rayon` and then aggregate the individual file "results" (which consist of the updated file contents and the "ultimately reported" violations for that file)